test-data/unit/check-based-intersection.test

```diff
@@ -0,0 +1,340 @@
+reveal_type(i.b)  # N: Revealed type is "str"
+reveal_type(i.c)  # E: "A & B" has no attribute "c"  [attr-defined] \
+                  # N: Revealed type is "Any (from error)"
+[builtins fixtures/tuple.pyi]
+
+
+[case testIntersectionAssignability]
+# flags: --disallow-redefinition
+from __future__ import annotations
+class A: pass
+class B: pass
+class AB(A, B): pass
+class D: pass
+
+ab: AB
+cd: AB & D
+ab_or_d: AB | D
+d: D
+i: A & B
+i = A()  # E: Incompatible types in assignment (expression has type "A", variable has type "A & B")  [assignment]
+ba: B & A = i
+i = ab
+i = d  # E: Incompatible types in assignment (expression has type "D", variable has type "A & B")  [assignment]
+i = ab_or_d  # E: Incompatible types in assignment (expression has type "AB | D", variable has type "A & B")  [assignment]
+[builtins fixtures/tuple.pyi]
+
+
+[case testIntersectionFromNarrowing]
+from __future__ import annotations
+class A: pass
+class B: pass
+
+i: A
+assert isinstance(i, B)
+reveal_type(i)  # N: Revealed type is "__main__.A & __main__.B"
+[builtins fixtures/tuple.pyi]
+
+
+[case testIntersectionWithBothCallable]
+from __future__ import annotations
+from typing import Callable
+class A:
+    def __call__(self, a: int): ...
+foo: A & Callable[[], int]
+reveal_type(foo())  # N: Revealed type is "int"
+foo("")  # E: No overload variant of "A & () -> int" matches argument type "str"  [call-overload] \
+         # N: Possible overload variants: \
+         # N:     def __call__(self, a: int) -> None \
+         # N:     def () -> int
+[builtins fixtures/tuple.pyi]
+
+
+[case testIntersectionOneCallable]
+from __future__ import annotations
+from typing import Callable
+class A:
+    pass
+foo: A & Callable[[], int]
+reveal_type(foo())  # N: Revealed type is "int"
+foo("")  # E: No overload variant of "A & () -> int" matches argument type "str"  [call-overload] \
+         # N: Possible overload variant: \
+         # N:     def () -> int
+[builtins fixtures/tuple.pyi]
+
+
+[case testIndexable]
+from __future__ import annotations
+
+class A:
+    def __getitem__(self, item: str): ...
+a: list[str] & A
+a[0]
+a["i"]
+a[None]  # E: No overload variant of "(int) -> str & (str) -> None" matches argument type "None"  [call-overload] \
+         # N: Possible overload variants: \
+         # N:     def __getitem__(self, int, /) -> str \
+         # N:     def __getitem__(self, str, /) -> None
+[builtins fixtures/list.pyi]
+
+
+[case testMultipleNarrowedIntersection]
+from __future__ import annotations
+class A: pass
+class B: pass
+class C: pass
+
+x: A
+reveal_type(x)  # N: Revealed type is "__main__.A"
+assert isinstance(x, B)
+reveal_type(x)  # N: Revealed type is "__main__.A & __main__.B"
+assert isinstance(x, C)
+reveal_type(x)  # N: Revealed type is "__main__.A & __main__.B & __main__.C"
+
+y: (A & B) | (A & C)
+assert isinstance(y, C)
+reveal_type(y)  # N: Revealed type is "__main__.A & __main__.B & __main__.C | __main__.A & __main__.C"
```

Comment:
this seems ambiguous, the revealed type should have the parentheses imo

Reply:
- #438